src/web/routes.rs

```diff
@@ -0,0 +1,257 @@
+        routes.internal_page(
+            &format!("/{}", redirect),
+            super::rustdoc::RustLangRedirector::new(redirect),
+        );
+    }
+
+    routes
+}
+
+pub(super) struct Routes {
+    get: Vec<(String, Box<dyn Handler>)>,
+    rustdoc_get: Vec<(String, Box<dyn Handler>)>,
+    page_prefixes: HashSet<String>,
+}
+
+impl Routes {
+    fn new() -> Self {
+        Self {
+            get: Vec::new(),
+            rustdoc_get: Vec::new(),
+            page_prefixes: HashSet::new(),
+        }
+    }
+
+    pub(super) fn page_prefixes(&self) -> HashSet<String> {
+        self.page_prefixes.clone()
+    }
+
+    pub(super) fn iron_router(mut self) -> Router {
+        let mut router = Router::new();
+        for (pattern, handler) in self.get.drain(..) {
+            router.get(&pattern, handler, calculate_id(&pattern));
+        }
+
+        // All rustdoc pages have the prefixes of other docs.rs pages blacklisted. This prevents,
+        // for example, a crate named "about" from hijacking /about/0.1.0/index.html.
+        let blacklist = self.page_prefixes();
+        for (pattern, handler) in self.rustdoc_get.drain(..) {
+            router.get(
+                &pattern,
+                BlockBlacklistedPrefixes::new(blacklist.clone(), handler),
+                calculate_id(&pattern),
+            );
+        }
+
+        router
+    }
+
+    /// A static resource is a normal page without any special behavior on the router side.
+    fn static_resource(&mut self, pattern: &str, handler: impl Handler) {
+        self.get.push((pattern.to_string(), Box::new(handler)));
+    }
+
+    /// Internal pages are docs.rs's own pages, instead of the documentation of a crate uploaded by
+    /// an user. The router adds these extra things when adding a new internal page:
+    ///
+    /// - The first component of the page's URL will be registered as a "page prefix". Page
+    /// prefixes are blacklisted by rustdoc pages, to prevent a crate named the same as a prefix
+    /// from hijacking docs.rs's own URLs.
+    ///
+    /// - If the page URL doesn't end with a slash, a redirect from the URL with the trailing slash
+    /// to the one without is automatically added.
+    fn internal_page(&mut self, pattern: &str, handler: impl Handler) {
+        self.get.push((pattern.to_string(), Box::new(handler)));
+
+        // Automatically add another route ending with / that redirects to the slash-less route.
+        if !pattern.ends_with('/') {
+            let pattern = format!("{}/", pattern);
+            self.get
+                .push((pattern.to_string(), Box::new(remove_slashes)));
+        }
+
+        // Register the prefix if it's not the home page and the first path component is not a
+        // pattern or a wildcard.
+        if pattern != "/" {
+            if let Some(first_component) = pattern.trim_matches('/').split('/').next() {
+                if !first_component.contains('*') && !first_component.starts_with(':') {
```

Comment:
I'm not quite sure what this is doing ... looking at the places `internal_page` is called it seems none of the prefixes meet this condition.

Reply:
This code is actually not needed right now, but it's a safeguard if in the future someone tries to add an internal page that starts with a wildcard.

Reply:
It's actually explained in the comment above, is that enough?

Reply:
Yes, I think that's fine.